crates/findex/src/pyo3/py_callbacks.rs

```diff
@@ -0,0 +1,320 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Display,
+};
+
+use cosmian_findex::{
+    parameters::{
+        DemScheme, KmacKey, BLOCK_LENGTH, DEM_KEY_LENGTH, KMAC_KEY_LENGTH, KWI_LENGTH,
+        MASTER_KEY_LENGTH, TABLE_WIDTH, UID_LENGTH,
+    },
+    CallbackError, EncryptedTable, FindexCallbacks, FindexCompact, FindexSearch, FindexUpsert,
+    IndexedValue as IndexedValueRust, Keyword, Location, Uid, UpsertData,
+};
+use pyo3::{
+    prelude::*,
+    types::{PyBytes, PyDict},
+};
+
+use crate::pyo3::{
+    py_api::InternalFindex,
+    py_structs::{Keyword as KeywordPy, Location as LocationPy},
+};
+
+#[derive(Debug)]
+pub enum FindexPyo3Error {
+    Callback(String),
+
+    ConversionError(String),
+}
+
+impl Display for FindexPyo3Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Callback(error) => write!(f, "callback error: {error}"),
+            Self::ConversionError(error) => write!(f, "conversion error: {error}"),
+        }
+    }
+}
+impl std::error::Error for FindexPyo3Error {}
+impl CallbackError for FindexPyo3Error {}
+
+impl FindexCallbacks<FindexPyo3Error, UID_LENGTH> for InternalFindex {
+    async fn progress(
+        &self,
+        results: &HashMap<Keyword, HashSet<IndexedValueRust>>,
+    ) -> Result<bool, FindexPyo3Error> {
+        Python::with_gil(|py| {
+            let py_results = results
+                .iter()
+                .map(|(keyword, locations)| {
+                    (
+                        KeywordPy(keyword.clone()),
+                        locations
+                            .iter()
+                            .map(|indexed_value| match indexed_value {
+                                IndexedValueRust::Location(location) => {
+                                    LocationPy(location.clone()).into_py(py)
+                                }
+                                IndexedValueRust::NextKeyword(keyword) => {
+                                    KeywordPy(keyword.clone()).into_py(py)
+                                }
+                            })
+                            .collect::<Vec<PyObject>>(),
+                    )
+                })
+                .collect::<HashMap<_, _>>();
+
+            let ret = self
+                .progress_callback
+                .call1(py, (py_results,))
+                .map_err(|e| FindexPyo3Error::Callback(format!("{e} (progress_callback)")))?;
+
+            ret.extract(py)
+                .map_err(|e| FindexPyo3Error::ConversionError(format!("{e} (progress_callback)")))
+        })
+    }
+
+    async fn fetch_all_entry_table_uids(
+        &self,
+    ) -> Result<HashSet<Uid<UID_LENGTH>>, FindexPyo3Error> {
+        Python::with_gil(|py| {
+            let results = self.fetch_all_entry_table_uids.call0(py).map_err(|e| {
+                FindexPyo3Error::Callback(format!("{e} (fetch_all_entry_table_uids)"))
+            })?;
```

Comment:
You may want to use the `pyo3_unwrap!()` marco (needs to implement `From<pyo3::PyErr>` for FindexPyo3Error):
```suggestion
            let results = pyo3_unwrap!(
                self.fetch_all_entry_table_uids.call0(py),
                "error fetching entry table"
            );
```

Reply:
There is two types of exceptions to raise: `FindexPyo3Error::Callback` and `FindexPyo3Error::ConversionError`.
This is why we did not implement `From<pyo3::PyErr>` before.